### ISSUE TITLE
Basic example doesn't display on iPhone 

### ISSUE THREAD

**Issue by Brun0oO** [hiukim/mind-ar-js#19]
Hi,

When I try to see the live demo of the basic example on my iPhone (6SPlus - iOS 14.4.1),
I stay on the waiting animation and if I take a look at the debug console, I got 
<img width="1215" alt="Capture d’écran 2021-03-30 à 12 06 33" src="https://user-images.githubusercontent.com/10347315/112972970-1ecc0e00-9151-11eb-8079-3251804fe2f0.png">

Any idea ?

**Comment by hiukim** [comment]
Probably something went wrong when running the webgl code.  Unfortunately, it's hard for me to debug without the device.

If you are interested, I can give some pointers and see if we can fix it together. But it requires you running the development build and tweak some codes and locate the errors.

**Comment by Brun0oO** [comment]
Yes, I think I can help you. I'm a developper, well, not a web one rather a
native application one ;o)
I have basic knowledge of debugging on my iOS device (using Safari).
So tell me everything !o)

Le mer. 31 mars 2021 à 01:57, HiuKim Yuen ***@***.***> a
écrit :

**Comment by hiukim** [comment]
@Brun0oO Thanks a lot. It would be greatly appreciated. 

First of all, you need to clone the repo and start development by running `npm run watch`. It will compile a development build of the library, i.e. `./dist-dev/mindar.js`. It will also watch for file changes and recompile continuously.

Then we can start debugging. I have just created a new debug example in the repo. `./examples/debug.html`. It's a minimal example that try to detect a simple image. You can run it as static html page. If everything runs successfully, you should see some console logs like this. I guess the first step is to use your desktop browser to verify you can get something like this.

![Screenshot 2021-03-31 at 11 04 01 AM](https://user-images.githubusercontent.com/459126/113190201-e55ad780-9210-11eb-8be3-e424eb2ec4e6.png)

After that, we can start debugging with the problematic device. We just need to locate the exact line that is causing the error. As you can see from debug.html, the entry point is `./src/controller.js`. Just a quick tip here: It's almost certain that the issue is related to the webgl code. After you dive into the codebase, you will find that there are codes that look like:
```
userCode:`
	void main() {
	  ivec2 coords = getOutputCoords();
	  int texR = coords[0];
	  int texC = coords[1];
	  vec2 uv = (vec2(texC, texR) + halfCR) / vec2(${width}.0, ${height}.0);
	  vec4 values = ${textureMethod}(A, uv);
	  setOutput((values.r + values.g + values.b) * 255.0 / 3.0);
	}
      `
```

Basically, these code are compiled and run in gpu. The issue is that they failed to compile. Let's try to locate the error and we'll go from there.

**Comment by Brun0oO** [comment]
I got some problems when I try to call the first "npm install" command in the main directory...  
I understand the problem is installing "node-sass@4.5.0" on my computer (macbook pro with macOS Catalina).
 
Here the complete log of "npm install" call (I truncated the long part which, in my opinion, does not give any relevant information):
```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'install' ]
2 info using npm@6.5.0-next.0
3 info using node@v11.6.0
4 verbose npm-session 70cb30c857581bab
5 silly install runPreinstallTopLevelLifecycles
6 silly preinstall mind-ar@0.3.2
7 info lifecycle mind-ar@0.3.2~preinstall: mind-ar@0.3.2
8 silly install loadCurrentTree
9 silly install readLocalPackageData
10 timing stage:loadCurrentTree Completed in 11ms
11 silly install loadIdealTree
12 silly install cloneCurrentTreeToIdealTree
13 timing stage:loadIdealTree:cloneCurrentTree Completed in 1ms
14 silly install loadShrinkwrap
15 timing stage:loadIdealTree:loadShrinkwrap Completed in 249ms
16 silly install loadAllDepsIntoIdealTree
17 timing stage:loadIdealTree:loadAllDepsIntoIdealTree Completed in 351ms
18 timing stage:loadIdealTree Completed in 715ms
19 silly currentTree mind-ar@0.3.2
20 silly idealTree mind-ar@0.3.2
20 silly idealTree ├── @discoveryjs/json-ext@0.5.2
20 silly idealTree ├── @msgpack/msgpack@1.12.2
20 silly idealTree ├── @npmcli/move-file@1.1.2
20 silly idealTree ├── @tensorflow/tfjs-backend-cpu@2.8.6
20 silly idealTree ├── @tensorflow/tfjs-backend-webgl@2.8.6
20 silly idealTree ├── @tensorflow/tfjs-converter@2.8.6
...
9967 silly saveTree │ └── webpack-sources@2.2.0
9967 silly saveTree └─┬ worker-loader@2.0.0
9967 silly saveTree   ├─┬ loader-utils@1.4.0
9967 silly saveTree   │ └── json5@1.0.1
9967 silly saveTree   └─┬ schema-utils@0.4.7
9967 silly saveTree     ├── ajv-keywords@3.4.1
9967 silly saveTree     └── ajv@6.12.2
9968 verbose stack Error: node-sass@5.0.0 postinstall: `node scripts/build.js`
9968 verbose stack Exit status 1
9968 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)
9968 verbose stack     at EventEmitter.emit (events.js:188:13)
9968 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
9968 verbose stack     at ChildProcess.emit (events.js:188:13)
9968 verbose stack     at maybeClose (internal/child_process.js:978:16)
9968 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:265:5)
9969 verbose pkgid node-sass@5.0.0
9970 verbose cwd /Users/bruno/GitHub/mind-ar-js
9971 verbose Darwin 19.6.0
9972 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
9973 verbose node v11.6.0
9974 verbose npm  v6.5.0-next.0
9975 error code ELIFECYCLE
9976 error errno 1
9977 error node-sass@5.0.0 postinstall: `node scripts/build.js`
9977 error Exit status 1
9978 error Failed at the node-sass@5.0.0 postinstall script.
9978 error This is probably not a problem with npm. There is likely additional logging output above.
9979 verbose exit [ 1, true ]
```

I've followed the advices of this post : https://stackoverflow.com/questions/48298361/npm-install-failed-at-the-node-sass4-5-0-postinstall-script without success.  

Here the paste of the console log of "npm i install node-sass@4.5.0" (once the previous advices has been applied) :
```
0 ✓ bruno@BRUNO ~/GitHub/mind-ar-js $ npm i node-sass@4.5.0
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported

> node-sass@4.5.0 install /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass
> node scripts/install.js

Downloading binary from https://github.com/sass/node-sass/releases/download/v4.5.0/darwin-x64-67_binding.node
Cannot download "https://github.com/sass/node-sass/releases/download/v4.5.0/darwin-x64-67_binding.node": 

HTTP error 404 Not Found

Hint: If github.com is not accessible in your location
      try setting a proxy via HTTP_PROXY, e.g. 

      export HTTP_PROXY=http://example.com:1234

or configure npm proxy via

      npm config set proxy http://example.com:8080

> node-sass@4.5.0 postinstall /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass
> node scripts/build.js

Building: /usr/local/bin/node /Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/bin/node-gyp.js rebuild --verbose --libsass_ext= --libsass_cflags= --libsass_ldflags= --libsass_library=
gyp info it worked if it ends with ok
gyp verb cli [ '/usr/local/bin/node',
gyp verb cli   '/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/bin/node-gyp.js',
gyp verb cli   'rebuild',
gyp verb cli   '--verbose',
gyp verb cli   '--libsass_ext=',
gyp verb cli   '--libsass_cflags=',
gyp verb cli   '--libsass_ldflags=',
gyp verb cli   '--libsass_library=' ]
gyp info using node-gyp@3.8.0
gyp info using node@11.6.0 | darwin | x64
gyp verb command rebuild []
gyp verb command clean []
gyp verb clean removing "build" directory
gyp verb command configure []
gyp verb check python checking for Python executable "python2" in the PATH
gyp verb `which` succeeded python2 /usr/local/bin/python2
gyp verb check python version `/usr/local/bin/python2 -c "import sys; print "2.7.15
gyp verb check python version .%s.%s" % sys.version_info[:3];"` returned: %j
gyp verb get node dir no --target version specified, falling back to host node version: 11.6.0
gyp verb command install [ '11.6.0' ]
gyp verb install input version string "11.6.0"
gyp verb install installing version: 11.6.0
gyp verb install --ensure was passed, so won't reinstall if already installed
gyp verb install version is already installed, need to check "installVersion"
gyp verb got "installVersion" 9
gyp verb needs "installVersion" 9
gyp verb install version is good
gyp verb get node dir target node version installed: 11.6.0
gyp verb build dir attempting to create "build" dir: /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass/build
gyp verb build dir "build" dir needed to be created? /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass/build
gyp verb build/config.gypi creating config file
gyp verb build/config.gypi writing out config file: /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass/build/config.gypi
gyp verb config.gypi checking for gypi file: /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass/config.gypi
gyp verb common.gypi checking for gypi file: /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass/common.gypi
gyp verb gyp gyp format was not specified; forcing "make"
gyp info spawn /usr/local/bin/python2
gyp info spawn args [ '/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/bruno/GitHub/mind-ar-js/node_modules/node-sass/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/bruno/.node-gyp/11.6.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/bruno/.node-gyp/11.6.0',
gyp info spawn args   '-Dnode_gyp_dir=/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/bruno/.node-gyp/11.6.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/bruno/GitHub/mind-ar-js/node_modules/node-sass',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
ERROR:root:code for hash md5 was not found.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type md5
ERROR:root:code for hash sha1 was not found.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha1
ERROR:root:code for hash sha224 was not found.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha224
ERROR:root:code for hash sha256 was not found.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha256
ERROR:root:code for hash sha384 was not found.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha384
ERROR:root:code for hash sha512 was not found.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha512
Traceback (most recent call last):
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/gyp_main.py", line 16, in <module>
    sys.exit(gyp.script_main())
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 545, in script_main
    return main(sys.argv[1:])
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 538, in main
    return gyp_main(args)
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 514, in gyp_main
    options.duplicate_basename_check)
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 98, in Load
    generator.CalculateVariables(default_variables, params)
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/pylib/gyp/generator/make.py", line 79, in CalculateVariables
    import gyp.generator.xcode as xcode_generator
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/pylib/gyp/generator/xcode.py", line 7, in <module>
    import gyp.xcodeproj_file
  File "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/gyp/pylib/gyp/xcodeproj_file.py", line 152, in <module>
    _new_sha1 = hashlib.sha1
AttributeError: 'module' object has no attribute 'sha1'
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/lib/configure.js:345:16)
gyp ERR! stack     at ChildProcess.emit (events.js:188:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:254:12)
gyp ERR! System Darwin 19.6.0
gyp ERR! command "/usr/local/bin/node" "/Users/bruno/GitHub/mind-ar-js/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
gyp ERR! cwd /Users/bruno/GitHub/mind-ar-js/node_modules/node-sass
gyp ERR! node -v v11.6.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
Build failed with error code: 1
npm WARN sass-loader@11.0.1 requires a peer of fibers@>= 3.1.0 but none is installed. You must install peer dependencies yourself.
npm WARN sass-loader@11.0.1 requires a peer of sass@^1.3.0 but none is installed. You must install peer dependencies yourself.
npm WARN worker-loader@2.0.0 requires a peer of webpack@^3.0.0 || ^4.0.0-alpha.0 || ^4.0.0 but none is installed. You must install peer dependencies yourself.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-sass@4.5.0 postinstall: `node scripts/build.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-sass@4.5.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/bruno/.npm/_logs/2021-04-02T07_07_05_352Z-debug.log
```
ƪ(ړײ)‎ƪ​​

**Comment by Brun0oO** [comment]
Hurrah! I managed to install by updating node to v15.13.0, clearing the cache and calling again "npm install". 
I continue following your instructions...

**Comment by Brun0oO** [comment]
On my desktop, I got this :
![Capture d’écran 2021-04-02 à 17 03 57](https://user-images.githubusercontent.com/10347315/113427599-9ca04b80-93d5-11eb-9b95-91c751fb3ceb.png)

On my iPhone, something strange appears, It's difficult to debug remotely with Safari. 
The debug window disappears quite often in the development menu on my desktop Safari, I don't know why yet. The rare times I was able to access the remote debug, I could see that mindar.js was not loaded in this specific case...

**Comment by hiukim** [comment]
@Brun0oO We are in the right track. Your desktop log is perfect. 

I remember that disappearing problem happens to me before. You can try this tool for remote debug: https://remotejs.com/

**Comment by Brun0oO** [comment]
Using _remotejs_, on my iOS device, I got this :
<img width="1508" alt="Capture d’écran 2021-04-06 à 08 48 29" src="https://user-images.githubusercontent.com/10347315/113670492-7d473e00-96b5-11eb-855e-6c96247f816b.png">

**Comment by Brun0oO** [comment]
New attempt with useful (perhaps!) informations :
<img width="1531" alt="Capture d’écran 2021-04-06 à 15 34 07" src="https://user-images.githubusercontent.com/10347315/113719398-dc28a980-96ed-11eb-9fb2-bfac42c7bb0a.png">

**Comment by hiukim** [comment]
> New attempt with useful (perhaps!) informations :
> <img alt="Capture d’écran 2021-04-06 à 15 34 07" width="1531" src="https://user-images.githubusercontent.com/10347315/113719398-dc28a980-96ed-11eb-9fb2-bfac42c7bb0a.png">

Just want to double check. Your iOS safari works on the example, right?

**Comment by Brun0oO** [comment]
Oups!, it works now, I don't understand what i did different!

**Comment by hiukim** [comment]
Does it work on debug.html or the basic example?

**Comment by Brun0oO** [comment]
I do not understand anything : now it doesn't work anymore on debug.html
I tried on example1.html without success, here the screenshots :
![IMG_D666F990FF79-1](https://user-images.githubusercontent.com/10347315/113781795-870e8700-9731-11eb-8dc5-a2c1196dbacd.jpeg)
<img width="1531" alt="Capture d’écran 2021-04-06 à 23 36 30" src="https://user-images.githubusercontent.com/10347315/113781855-9beb1a80-9731-11eb-9ed2-a9a1a6f2f6df.png">


I noticed I have the same error as for displaying debug.html
but I didn't dream (at least I think), I (only) saw the debug.html page with the camera display once.
What is the expected behaviour of the debug.html page, should I see the camera if all goes well?

**Comment by hiukim** [comment]
You should see an empty page in debug.html.  Just the console.log. Anyway, I think we were expecting error, so that's fine.


Can you quickly try one thing? In the `src/image-target/input-loader.js` make it use the old method `loadInput` method

Basically, just rename `_loadInput` to `loadInput` and `loadInput` to `_loadInput`

https://github.com/hiukim/mind-ar-js/blob/ba310848f10acf3941bdccd949b3120b14ddda96/src/image-target/input-loader.js#L29-L48

**Comment by Brun0oO** [comment]
I've reversed _loadInput and loadInput.
I took the opportunity to add some traces (logs like ">0", ">1", ...)  :
```javascript
...
class InputLoader {
  constructor(width, height) {
    console.log(">0");
    this.width = width;
    this.height = height;
    this.texShape = [height, width];

    const context = document.createElement('canvas').getContext('2d');
    context.canvas.width = width;
    context.canvas.height = height;
    this.context = context;

    this.program = this.buildProgram(width, height);

    const backend = tf.backend();
    console.log(">1");
    console.log(backend);
    this.tempPixelHandle = backend.makeTensorInfo(this.texShape, 'int32');
    // warning!!!
    // usage type should be TextureUsage.PIXELS, but tfjs didn't export this enum type, so we hard-coded 2 here 
    //   i.e. backend.texData.get(tempPixelHandle.dataId).usage = TextureUsage.PIXELS;
    console.log(">2");
    backend.texData.get(this.tempPixelHandle.dataId).usage = 2;
    console.log(">3");
  }

  // old method
  loadInput(input) {
    console.log(">4");
    return tf.tidy(() => {
      let inputImage = tf.browser.fromPixels(input);
      inputImage = inputImage.mean(2);
      return inputImage;
    });
  }

  // input is instance of HTMLVideoElement or HTMLImageElement
  _loadInput(input) {
    this.context.drawImage(input, 0, 0, this.width, this.height);

    const backend = tf.backend();
    backend.gpgpu.uploadPixelDataToTexture(backend.getTexture(this.tempPixelHandle.dataId), this.context.canvas);

    const res = backend.compileAndRun(this.program, [this.tempPixelHandle], 'float32');
    //backend.disposeData(tempPixelHandle.dataId);
    return res;
  }
 ...
```
Here the results :
<img width="1487" alt="Capture d’écran 2021-04-07 à 10 33 12" src="https://user-images.githubusercontent.com/10347315/113836117-e483f180-978c-11eb-95f6-1d18465f7fe6.png">

I've tried to inspect the backend variable but I did not succeed :
* Using console.log(JSON.stringify(backend, null, 2); I got a "JSON.stringify cannot serialize cyclic structures." message.
* Using console.dir(backend); I got ... nothing !

It is as if the tensorflow backend was not yet initialized when the InputLoader constructor was called... 
Do we have to wait until the tensorflow engine is ready?
I noticed that the loadInput function was not called when the exception is raised.

**Comment by hiukim** [comment]
ah.. let's try comment out these part of code as well. we don't need them for the old loadInput 

```
/*
this.program = this.buildProgram(width, height);
    const backend = tf.backend();
    console.log(">1");
    console.log(backend);
    this.tempPixelHandle = backend.makeTensorInfo(this.texShape, 'int32');
    // warning!!!
    // usage type should be TextureUsage.PIXELS, but tfjs didn't export this enum type, so we hard-coded 2 here 
    //   i.e. backend.texData.get(tempPixelHandle.dataId).usage = TextureUsage.PIXELS;
    console.log(">2");
    backend.texData.get(this.tempPixelHandle.dataId).usage = 2;
    console.log(">3");
*/
```

**Comment by Brun0oO** [comment]
We are going forward ;o) now, loadInput is called, new screenshot :
<img width="1531" alt="Capture d’écran 2021-04-07 à 20 11 44" src="https://user-images.githubusercontent.com/10347315/113914234-9ea34980-97dd-11eb-93d3-b74caf62fbbe.png">

**Comment by hiukim** [comment]
seems like the tensorflow webgl backend has failed to initialize, which is also indicated in the Warning. 

Probably need to dig deeper into that.
